File: static/gallery_res/demo14/demo14.cpp
/*
[A⁻¹+A⁻¹BS⁻¹BᵀA⁻¹   -A⁻¹BS⁻¹
    -S⁻¹BᵀA⁻¹           S⁻¹]

where

A ∈ ℝ^(4×4) 
B ∈ ℝ^(4×4) 
S ∈ ℝ^(4×4) 
*/
#include <Eigen/Core>
#include <Eigen/Dense>
#include <Eigen/Sparse>
#include <iostream>
#include <set>

Eigen::Matrix<double, 8, 8> demo14(
    const Eigen::Matrix<double, 4, 4> & A,
    const Eigen::Matrix<double, 4, 4> & B,
    const Eigen::Matrix<double, 4, 4> & S)
{
    Eigen::Matrix<double, 8, 8> _ret_0;
    _ret_0 << A.inverse() + A.inverse() * B * S.inverse() * B.transpose() * A.inverse(), -A.inverse() * B * S.inverse(),
    -S.inverse() * B.transpose() * A.inverse(), S.inverse();
    Eigen::Matrix<double, 8, 8> ret = _ret_0;
    return ret;
}


void generateRandomData(Eigen::Matrix<double, 4, 4> & A,
    Eigen::Matrix<double, 4, 4> & B,
    Eigen::Matrix<double, 4, 4> & S)
{
    A = Eigen::MatrixXd::Random(4, 4);
    B = Eigen::MatrixXd::Random(4, 4);
    S = Eigen::MatrixXd::Random(4, 4);
}


int main(int argc, char *argv[])
{
    srand((int)time(NULL));
    Eigen::Matrix<double, 4, 4> A;
    Eigen::Matrix<double, 4, 4> B;
    Eigen::Matrix<double, 4, 4> S;
    generateRandomData(A, B, S);
    Eigen::Matrix<double, 8, 8> func_value = demo14(A, B, S);
    std::cout<<"func_value:\n"<<func_value<<std::endl;
    return 0;
}

File: static/gallery_res/demo14/demo14.py
"""
[A⁻¹+A⁻¹BS⁻¹BᵀA⁻¹   -A⁻¹BS⁻¹
    -S⁻¹BᵀA⁻¹           S⁻¹]

where

A: ℝ^(4×4) 
B: ℝ^(4×4) 
S: ℝ^(4×4) 
"""
import numpy as np
import scipy
import scipy.linalg
from scipy import sparse
from scipy.integrate import quad
from scipy.optimize import minimize


def demo14(A, B, S):
    """
    :param :A : ℝ^(4×4)
    :param :B : ℝ^(4×4)
    :param :S : ℝ^(4×4)
    """
    A = np.asarray(A, dtype=np.float64)
    B = np.asarray(B, dtype=np.float64)
    S = np.asarray(S, dtype=np.float64)

    assert A.shape == (4, 4)
    assert B.shape == (4, 4)
    assert S.shape == (4, 4)

    _ret_0 = np.block([[np.linalg.inv(A) + np.linalg.inv(A) @ B @ np.linalg.inv(S) @ B.T @ np.linalg.inv(A), -np.linalg.inv(A) @ B @ np.linalg.inv(S)], [-np.linalg.inv(S) @ B.T @ np.linalg.inv(A), np.linalg.inv(S)]])
    ret = _ret_0
    return ret


def generateRandomData():
    A = np.random.randn(4, 4)
    B = np.random.randn(4, 4)
    S = np.random.randn(4, 4)
    return A, B, S


if __name__ == '__main__':
    A, B, S = generateRandomData()
    print("A:", A)
    print("B:", B)
    print("S:", S)
    func_value = demo14(A, B, S)
    print("func_value: ", func_value)